src/read/read_cache.rs

```diff
@@ -0,0 +1,120 @@
+use std::boxed::Box;
+use std::cell::RefCell;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::io::{Read, Seek, SeekFrom};
+use std::mem;
+
+use crate::read::ReadRef;
+
+/// An implementation of `ReadRef` for data in a stream that implements
+/// `Read + Seek`.
+///
+/// Contains a cache of read-only blocks of data, allowing references to
+/// them to be returned. Entries in the cache are never removed.
+/// Entries are keyed on the offset and size of the read.
+/// Currently overlapping reads are considered separate reads.
+#[derive(Debug)]
+pub struct ReadCache<R: Read + Seek> {
+    cache: RefCell<ReadCacheInternal<R>>,
+}
+
+#[derive(Debug)]
+struct ReadCacheInternal<R: Read + Seek> {
+    read: R,
+    bufs: HashMap<(u64, u64), Box<[u8]>>,
+}
+
+impl<R: Read + Seek> ReadCache<R> {
```

Comment:
It may make sense to have a

```rust
pub fn clear(&mut self) { ... }
```

method here. The `&mut self` would be proof that any existing `ReadRef` users were done with this cache, but would allow reusing the internal `HashMap`'s storage (and maybe, with a bit more work, there could be a free list of available buffers to reuse, but maybe not worth the complexity and also maybe not even a perf win over just freeing to the global allocator and then allocating new buffers again).